node/core/approval-voting/src/lib.rs

```diff
@@ -774,6 +793,162 @@ where
 	Ok(())
 }
 
+#[derive(Eq, PartialEq)]
+enum RecoveryState {
+	Pending,
+	Complete,
+}
+
+impl RecoveryState {
+	fn peek(&self) -> bool {
+		*self == RecoveryState::Complete
+	}
+
+	fn complete(&mut self) -> bool {
+		let complete = self.peek();
+		*self = RecoveryState::Complete;
+		complete
+	}
+}
+
+// Determines the set of assigned but unapproved candidates that require resumption of the
+// approval process by the local node. For each such, this method generates a `LaunchApproval`
+// action so that the approvals can continue to make progress after a restart.
+async fn handle_startup(
+	overlay_db: &mut OverlayedBackend<'_, impl Backend>,
+	state: &mut State,
+) -> SubsystemResult<Vec<(BlockNumber, Action)>> {
+	let all_candidates = overlay_db.load_all_candidates().map_err(|e| {
+		tracing::error!(target: LOG_TARGET, "Failed initial load of active candidates: {:?}", e);
+		e
+	})?;
+
+	let mut actions = Vec::new();
+	for candidate in all_candidates {
+		let mut assigned_but_unapproved_candidates = Vec::new();
+		let session = candidate.session;
+		let info = match state.session_info(session) {
+			Some(info) => info,
+			None => {
+				tracing::debug!(target: LOG_TARGET, "Missing session info for {:?}", session);
+				continue
+			},
+		};
+
+		// Compute the local validator indexes, if any, for this session.
+		let local_validators: Vec<_> = info
+			.validators
+			.iter()
+			.enumerate()
+			.filter(|(_, validator)| {
+				state
+					.keystore
+					.key_pair::<ValidatorPair>(validator)
+					.ok()
+					.map_or(false, |v| v.is_some())
+			})
+			.map(|(index, _)| ValidatorIndex(index as _))
+			.collect();
+
+		// Compute the local assignment indexes, if any, for this session.
+		let local_assigners: Vec<_> = info
+			.assignment_keys
+			.iter()
+			.enumerate()
+			.filter_map(|(index, assigner)| {
+				state
+					.keystore
+					.key_pair::<AssignmentPair>(assigner)
+					.ok()
+					.map(|_| ValidatorIndex(index as _))
+			})
+			.collect();
+
+		let is_session_validator = !local_validators.is_empty();
+		let is_session_assigner = !local_assigners.is_empty();
+
+		// Determine whether or not the local node controls session validation and assignment. If
+		// not, skip all candidates in this session.
+		let controls_validation_and_assignment = is_session_validator && is_session_assigner;
+		if !controls_validation_and_assignment {
+			continue
+		}
+
+		// Determine the candidates in this session that need to have their approvals resumed.
```

Comment:
Comment isn't accurate. This loop is actually finding all the locally-controlled validators on this specific candidate that need to have their approval resumed.